kolibri/core/public/api.py

```diff
@@ -220,134 +216,135 @@ def get_public_file_checksums(request, version):
 
-        if not FacilityUser.objects.filter(id=user).exists():
+        if not FacilityUser.objects.filter(id=user_id).exists():
             content = "This user is not registered in any of this server facilities"
-            return Response(content, status=status.HTTP_404_NOT_FOUND)
+            raise exceptions.NotFound(content)
+
+        with transaction.atomic():
+            queue_object, created = SyncQueue.objects.get_or_create(
+                id=pk,
+                defaults=dict(
+                    user_id=user_id,
+                    instance_id=instance_id,
+                    status=SyncQueueStatus.Queued,
+                ),
+            )
 
-        # first, ensure no expired devices are in the queue
-        SyncQueue.clean_stale()
+        if not created:
+            if (
+                queue_object.user_id != user_id
+                or queue_object.instance_id != instance_id
+            ):
+                content = "A queue object exists with different user and instance ids"
+                raise self.validation_error(content, status.HTTP_409_CONFLICT)
 
-        # Calculate the total size of the queue to scale things
-        total_queue_size = SyncQueue.objects.count()
+            queue_object.status = SyncQueueStatus.Queued
+            queue_object.datetime = timezone.now()
+            queue_object.set_next_attempt(HANDSHAKING_TIME)
 
-        # Scale the sync_interval we send to clients based on the total number of
-        # queued single user sync requests.
-        # Make sure it is no longer than half the time by which we measure 'recently synced'
-        # to make sure we are never letting syncing drift too far.
-        sync_interval = min(
-            OPTIONS["Deployment"]["SYNC_INTERVAL"] * (total_queue_size + 1),
-            DELAYED_SYNC / 2,
-        )
+        if queue_object.last_sync is None:
+            last_sync = soud.get_last_successful_sync(instance_id)
 
-        if pk is not None:
-            queue_object = SyncQueue.objects.filter(id=pk).first()
-        else:
-            queue_object = SyncQueue.objects.filter(
-                user_id=user, instance_id=instance
-            ).first()
-
-        # Default the position to the total queue size, so that
-        # if the id does not exist, send them to the back of the queue
-        pos = total_queue_size
-        if queue_object is not None:
-            if queue_object.user_id != user:
-                return Response(
-                    "Queue did not match user in request",
-                    status=status.HTTP_400_BAD_REQUEST,
-                )
-            if queue_object.instance_id != instance:
-                return Response(
-                    "Queue did not match instance in request",
-                    status=status.HTTP_400_BAD_REQUEST,
+            if last_sync:
+                queue_object.last_sync = last_sync.timestamp()
```

Comment:
`timestamp` function was introduced in python 3.3 (https://docs.python.org/3/library/datetime.html#datetime.datetime.timestamp).
Needs to be replaced by
`time.mktime(last_sync.timetuple())`
for 2.7 compatibility